### ISSUE TITLE
Possibility to directly reference the production version of a model

### ISSUE THREAD

**Issue by kaifox** [jmad/jmad-modelpack-service#4]
e.g. through repo-url + package-name + name 

somehow like:

`{'https://gitlab.cern.ch/jmad-modelpacks-cern', 'jmad-modelpack-lhc', 'LHC'}`

Shall we call it 'groupId' and 'artifactId'?

**Comment by michi42** [comment]
While integrating the local-file connector I was thinking of something similar. I'm wondering, should we maybe use URLs throughout as unique identifiers to modelpack repositories, model packs, and even individual models? To me that looks like the "natural" way ;-)
Of course the URL structure would be connector-dependent, for instance to refer to an internal model one could use something like 'jmad://internal#SPS'. For a model on gitlab, it could look like 'gitlab+https://gitlab.cern.ch/jmad-modelpacks-cern/jmad-modelpack-lhc#LHC' for the latest release, or even 'gitlab+https://gitlab.cern.ch/jmad-modelpacks-cern/jmad-modelpack-lhc/releases/v2018.1#LHC' for a particular one. For the local file system, we would have something like 'file:///home/michi/jmad-models/jmad-modelpack-lhc#LHC'.

A similar concept could be used to refer to modelpack variants (just without the hash-part), or to modelpack repositories (just put the base URL). It would add a little more logic to the connectors to parse these URLs, on the other hand, the general format would be standardized, connector agnostic, and probably already known to many users.

This would also allow sharing and adding repositories/model packs e.g. in private repos at runtime with no big overhead - just have an "add" button which allows you to copy-paste an URL.

I had something simlar on the python side already, but only for modelpack repositories.

**Comment by michi42** [comment]
Some URL proposals after discussing with Andrea:

gitlab+https://gitlab.cern.ch/jmad-modelpacks-cern/jmad-modelpack-lhc?release=v2018.1#LHC
gitlab+https://gitlab.cern.ch/jmad-modelpacks-cern/jmad-modelpack-lhc?tag=v2018.1#LHC
gitlab+https://gitlab.cern.ch/jmad-modelpacks-cern/jmad-modelpack-lhc?branch=v2018.1#LHC

... or alternatively:
gitlab://default-cern/jmad-mocelpack-lhc#LHC

classpath://#LHC

file:///User/michi/models#LHC

**Comment by kaifox** [comment]
I must admit, although I was a bit skeptical in the beginning, I agree now. Indeed this looks the natural way in our netified world ;-)

I like your proposals. Indeed it would be nice to have simple urls for the default ones and be able to adress more specific ones also ....

... One could even think further: Could we even address a specific sequence + optics like this and initialize a model in a one liner?

something like:

```
JMadModel model = service.create("gitlab+https://gitlab.cern.ch/jmad-modelpacks-cern/jmad-modelpack-lhc?release=v2018.1&optics=XXXXXX&sequence=lhcb1#LHC");
```

A well defined syntax would indeed appreciated and would be nice to have the parsing mechanism extracted to a common utility class..

**Comment by kaifox** [comment]
BTW: Not sure if I like the #LHC at the end ... could this also be just another path entry (like "/LHC")? ... Then all the parameters would come afterwards...

**Comment by michi42** [comment]
The idea behind the fragment-identifier part (#LHC etc) was that the rest of the URL is parsed by the connector (internal/classpath, local file, gitlab, or possible more in the future) to identify a modelpack-variant. This may involve making requests to remote servers, downloading that modelpack-variant from gitlab etc.

Then the fragment-identifier is parsed by common code, after the connector has returned, to identify one particular model in that modelpack-variant, e.g. LHC. This has nothing to do anymore with where that modelpack-variant came from. Also the optics, sequence etc. (which I agree would be nice to have) should go there, following the concept of https://en.wikipedia.org/wiki/Fragment_identifier

**Comment by michi42** [comment]
I would like to bring this back up, as we start having the situation in multiple places where in Run 2 we used to refer to a hard-coded model from an internal modelpack, and where we have all the information to auto-select the "right" model, e.g. in LSA trim simulation.

I think, in general, we should foresee a URL like the following
```
<connector-identifier>://<connector-specific-part>#<model&sequence&optic>
```
The idea is:
- `<connector-identifier>`, the "protocol", **identifies a modelpack-connector** to use. It is parsed by the modelpack service.
- `<connector-specific-part>` **identifies a particular model pack**. It is parsed by the connector, e.g. the GitLab connector. It is up to the connector how this part looks like. For some connectors like the classpath one, it could be empty.
- `<model&sequence&optic>` **identifies a particular model** within the model pack, and optionally sets up a sequence and/or optic to use. It is parsed by JMad or the modelpack service (to be seen). It is common for any connector.

For GitLab, I checked again how pip handles this as they have a similar concept, one can "[https://pip.pypa.io/en/stable/reference/pip_install/#vcs-support](pip install git+https://...").
Maybe it's best to simply follow the same convention:
gitlab+https://gitlab.cern.ch/acc-models/acc-models-lhc#...
(for master)
gitlab+https://gitlab.cern.ch/acc-models/acc-models-lhc@v2020.1#...
(from tag/release/branch - they share a namespace in git)

The fragment "#..." part could be the following:
- ...#LHC(just the name of the model)
- ...#LHC&seq=lhcb1 (model and sequence)
- ...#LHC&optic=R2018a_... (model and optic)
- ...#LHC&seq=lhcb1&optic=R2018a_... (everything)

The idea would be, that when uploading an optic (which should always be a tagged version), we would generate the GitLab URL in the optic uploader and store it in LSA as meta-information, so we would always have all the information we need to re-produce it in JMad.
